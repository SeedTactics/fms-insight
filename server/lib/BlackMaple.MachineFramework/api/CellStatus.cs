/* Copyright (c) 2023, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#nullable enable

using System;
using System.Runtime.Serialization;
using System.Collections.Immutable;
using Germinate;

namespace BlackMaple.MachineFramework
{
  [DataContract]
  public record QueueSize
  {
    //once an output queue grows to this size, stop unloading parts
    //and keep them in the buffer inside the cell
    [DataMember(IsRequired = false, EmitDefaultValue = false)]
    public int? MaxSizeBeforeStopUnloading { get; init; }
  }

  [DataContract, Draftable]
  public record CurrentStatus
  {
    [DataMember(IsRequired = true)]
    public required DateTime TimeOfCurrentStatusUTC { get; init; }

    [DataMember(Name = "Jobs", IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.ActiveJob> Jobs { get; init; }

    [DataMember(Name = "Pallets", IsRequired = true)]
    public required ImmutableDictionary<string, PalletStatus> Pallets { get; init; }

    [DataMember(Name = "Material", IsRequired = true)]
    public required ImmutableList<InProcessMaterial> Material { get; init; }

    [DataMember(Name = "Alarms", IsRequired = true)]
    public required ImmutableList<string> Alarms { get; init; }

    [DataMember(Name = "Queues", IsRequired = true)]
    public required ImmutableDictionary<string, QueueSize> QueueSizes { get; init; }

    // The following is only filled in if the machines move to the load station
    // instead of the pallets moving to the machine
    [DataMember(Name = "MachineLocations", IsRequired = false, EmitDefaultValue = false)]
    public ImmutableList<MachineLocation>? MachineLocations { get; init; }

    public static CurrentStatus operator %(CurrentStatus s, Action<ICurrentStatusDraft> f) => s.Produce(f);
  }

  [DataContract]
  public record JobAndDecrementQuantity
  {
    [DataMember(IsRequired = true)]
    public required long DecrementId { get; init; }

    [DataMember(IsRequired = true)]
    public required string JobUnique { get; init; }

    [DataMember(IsRequired = true)]
    public required DateTime TimeUTC { get; init; }

    [DataMember(IsRequired = true)]
    public required string Part { get; init; }

    [DataMember(IsRequired = true)]
    public required int Quantity { get; init; }
  }

  [DataContract]
  public record HistoricData
  {
    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<SimulatedStationUtilization> StationUse { get; init; }
  }

  [DataContract]
  public record PlannedSchedule
  {
    [DataMember(IsRequired = true)]
    public required string LatestScheduleId { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, int> ExtraParts { get; init; }

    [DataMember(IsRequired = false)]
    public ImmutableList<Workorder>? CurrentUnfilledWorkorders { get; init; }
  }
}
