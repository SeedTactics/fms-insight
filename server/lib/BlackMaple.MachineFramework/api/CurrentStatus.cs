/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;

namespace BlackMaple.MachineWatchInterface
{
  ///Stores what is currently happening to a piece of material.
  [Serializable, DataContract]
  public class InProcessMaterialAction
  {
    // This should be a sum type, and while C# sum types can work with some helper code it doesn't work
    // well for serialization.

    [Serializable, DataContract]
    public enum ActionType
    {
      Waiting = 0,
      Loading,
      UnloadToInProcess, // unload, but keep the material around because more processes must be machined
      UnloadToCompletedMaterial, // unload and the material has been completed
      Machining
    }
    [DataMember(IsRequired = true)] public ActionType Type { get; set; }

    // If Type = Loading
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string LoadOntoPallet { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public int? LoadOntoFace { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public int? ProcessAfterLoad { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public int? PathAfterLoad { get; set; }

    //If Type = UnloadToInProcess
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string UnloadIntoQueue { get; set; }

    //If Type = Loading or UnloadToInProcess or UnloadToCompletedMaterial
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public TimeSpan? ElapsedLoadUnloadTime { get; set; }

    // If Type = Machining
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string Program { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public TimeSpan? ElapsedMachiningTime { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public TimeSpan? ExpectedRemainingMachiningTime { get; set; }
  }

  ///Stores the current location of a piece of material.  If a transfer operation is currently in process
  ///(such as unloading), the location will store the previous location and the action will store the new location.
  [Serializable, DataContract]
  public class InProcessMaterialLocation
  {
    //Again, this should be a sum type.
    [Serializable, DataContract]
    public enum LocType
    {
      Free = 0,
      OnPallet,
      InQueue,
    }
    [DataMember(IsRequired = true)] public LocType Type { get; set; }

    //If Type == OnPallet
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string Pallet { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public int? Face { get; set; }

    //If Type == InQueue
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string CurrentQueue { get; set; }

    //If Type == InQueue or Type == Free
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public int? QueuePosition { get; set; }
  }

  //Stores information about a piece of material, where it is, and what is happening to it.
  [Serializable, DataContract]
  public class InProcessMaterial
  {
    // Information about the material
    [DataMember(IsRequired = true)] public long MaterialID { get; set; }
    [DataMember(IsRequired = true)] public string JobUnique { get; set; }
    [DataMember(IsRequired = true)] public string PartName { get; set; }
    [DataMember(IsRequired = true)] public int Process { get; set; }  // When in a queue, the process is the last completed process
    [DataMember(IsRequired = true)] public int Path { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string Serial { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string WorkorderId { get; set; }
    [DataMember(IsRequired = true)] public List<string> SignaledInspections { get; set; } = new List<string>();

    // 0-based index into the JobPlan.MachiningStops array for the last completed stop.  Null or negative values
    // indicate no machining stops have yet completed.
    [DataMember(IsRequired = false)] public int? LastCompletedMachiningRouteStopIndex { get; set; }

    // Where is the material?
    [DataMember(IsRequired = true)] public InProcessMaterialLocation Location { get; set; }

    // What is currently happening to the material?
    [DataMember(IsRequired = true)] public InProcessMaterialAction Action { get; set; }
  }

  [Serializable, DataContract]
  public enum PalletLocationEnum
  {
    [EnumMember] LoadUnload,
    [EnumMember] Machine,
    [EnumMember] MachineQueue,
    [EnumMember] Buffer,
    [EnumMember] Cart
  }

  [Serializable, DataContract]
  public struct PalletLocation : IComparable
  {

    [DataMember(Name = "loc", IsRequired = true)]
    public PalletLocationEnum Location { get; set; }

    [DataMember(Name = "group", IsRequired = true)]
    public string StationGroup { get; set; }

    [DataMember(Name = "num", IsRequired = true)]
    public int Num { get; set; }

    public PalletLocation(PalletLocationEnum l, string group)
    {
      Location = l;
      StationGroup = group;
      Num = 1;
    }

    public PalletLocation(PalletLocationEnum l, string group, int n)
    {
      Location = l;
      StationGroup = group;
      Num = n;
    }

    public int CompareTo(object obj)
    {
      PalletLocation other = (PalletLocation)obj;
      int cmp = Location.CompareTo(other.Location);
      if (cmp < 0)
        return -1;
      if (cmp > 0)
        return 1;
      cmp = StationGroup.CompareTo(other.StationGroup);
      if (cmp < 0)
        return -1;
      if (cmp > 0)
        return 1;
      return Num.CompareTo(other.Num);
    }
  }

  [Serializable, DataContract]
  public class PalletStatus
  {
    [DataMember(IsRequired = true)] public string Pallet { get; set; }
    [DataMember(IsRequired = true)] public string FixtureOnPallet { get; set; }
    [DataMember(IsRequired = true)] public bool OnHold { get; set; }
    [DataMember(IsRequired = true)] public PalletLocation CurrentPalletLocation { get; set; }

    // If the pallet is at a load station and a new fixture should be loaded, this is filled in.
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public string NewFixture { get; set; }

    // num faces on new fixture, or current fixture if no change
    [DataMember(IsRequired = true)] public int NumFaces { get; set; }

    //If CurrentPalletLocation is Cart, the following two fields will be filled in.
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public PalletLocation? TargetLocation { get; set; }
    [DataMember(IsRequired = false, EmitDefaultValue = false)] public decimal? PercentMoveCompleted { get; set; }
  }

  [Serializable, DataContract]
  public class InProcessJobDecrement
  {
    [DataMember(IsRequired = true)] public long DecrementId { get; set; }
    [DataMember(IsRequired = true)] public DateTime TimeUTC { get; set; }
    [DataMember(IsRequired = true)] public int Quantity { get; set; }
  }

  [SerializableAttribute, DataContract]
  public class InProcessJob : JobPlan
  {
    public int GetCompleted(int process, int path)
    {
      if (process >= 1 && process <= NumProcesses && path >= 1 && path <= GetNumPaths(process))
      {
        return _completed[process - 1][path - 1];
      }
      else
      {
        throw new IndexOutOfRangeException("Invalid process or path number");
      }
    }
    public void SetCompleted(int process, int path, int comp)
    {
      if (process >= 1 && process <= NumProcesses && path >= 1 && path <= GetNumPaths(process))
      {
        _completed[process - 1][path - 1] = comp;
      }
      else
      {
        throw new IndexOutOfRangeException("Invalid process or path number");
      }
    }
    public void AdjustCompleted(int process, int path, Func<int, int> f)
    {
      if (process >= 1 && process <= NumProcesses && path >= 1 && path <= GetNumPaths(process))
      {
        _completed[process - 1][path - 1] = f(_completed[process - 1][path - 1]);
      }
      else
      {
        throw new IndexOutOfRangeException("Invalid process or path number");
      }
    }

    public long GetPrecedence(int process, int path)
    {
      if (_precedence == null) return -1;
      if (process >= 1 && process <= NumProcesses && path >= 1 && path <= GetNumPaths(process))
      {
        return _precedence[process - 1][path - 1];
      }
      else
      {
        throw new IndexOutOfRangeException("Invalid process or path number");
      }
    }

    public void SetPrecedence(int process, int path, long precedence)
    {
      if (_precedence == null)
      {
        _precedence = new long[base.NumProcesses][];
        for (int proc = 1; proc <= base.NumProcesses; proc++)
        {
          _precedence[proc - 1] = Enumerable.Repeat(-1L, base.GetNumPaths(proc)).ToArray();
        }
      }
      if (process >= 1 && process <= NumProcesses && path >= 1 && path <= GetNumPaths(process))
      {
        _precedence[process - 1][path - 1] = precedence;
      }
      else
      {
        throw new IndexOutOfRangeException("Invalid process or path number");
      }
    }

    public IList<InProcessJobDecrement> Decrements
    {
      get
      {
        if (_decrements == null)
          _decrements = new List<InProcessJobDecrement>();
        return _decrements;
      }
    }

    public InProcessJob(string unique, int numProc, int[] numPaths = null) : base(unique, numProc, numPaths)
    {
      _completed = new int[base.NumProcesses][];
      for (int proc = 1; proc <= base.NumProcesses; proc++)
      {
        _completed[proc - 1] = Enumerable.Repeat(0, base.GetNumPaths(proc)).ToArray();
      }
    }
    public InProcessJob(JobPlan job) : base(job)
    {
      _completed = new int[base.NumProcesses][];
      for (int proc = 1; proc <= base.NumProcesses; proc++)
      {
        _completed[proc - 1] = Enumerable.Repeat(0, base.GetNumPaths(proc)).ToArray();
      }
    }

    private InProcessJob() { } //for json deserialization

    [DataMember(Name = "Completed", IsRequired = false)] private int[][] _completed;

    [DataMember(Name = "Decrements", IsRequired = false), OptionalField] private List<InProcessJobDecrement> _decrements;

    // a number reflecting the order in which the cell controller will consider the processes and paths for activation.
    // lower numbers come first, while -1 means no-data.
    [DataMember(Name = "Precedence", IsRequired = false), OptionalField] private long[][] _precedence;
  }

  [SerializableAttribute, DataContract]
  public class CurrentStatus
  {
    [DataMember(IsRequired = true)] public DateTime TimeOfCurrentStatusUTC { get; set; }
    public IDictionary<string, InProcessJob> Jobs => _jobs;
    public IDictionary<string, PalletStatus> Pallets => _pals;
    public IList<InProcessMaterial> Material => _material;
    public IList<string> Alarms => _alarms;
    public Dictionary<string, QueueSize> QueueSizes => _queues;

    public CurrentStatus(DateTime? nowUTC = null)
    {
      TimeOfCurrentStatusUTC = nowUTC ?? DateTime.UtcNow;
      _jobs = new Dictionary<string, InProcessJob>();
      _pals = new Dictionary<string, PalletStatus>();
      _material = new List<InProcessMaterial>();
      _alarms = new List<string>();
      _queues = new Dictionary<string, QueueSize>();
    }

    [DataMember(Name = "Jobs", IsRequired = true)]
    private Dictionary<string, InProcessJob> _jobs;

    [DataMember(Name = "Pallets", IsRequired = true)]
    private Dictionary<string, PalletStatus> _pals;

    [DataMember(Name = "Material", IsRequired = true)]
    private List<InProcessMaterial> _material;

    [DataMember(Name = "Alarms", IsRequired = true)]
    private List<string> _alarms;

    [DataMember(Name = "Queues", IsRequired = true)]
    private Dictionary<string, QueueSize> _queues;
  }

  [Serializable, DataContract]
  public struct JobAndDecrementQuantity
  {
    [DataMember(IsRequired = true)] public long DecrementId { get; set; }
    [DataMember(IsRequired = true)] public string JobUnique { get; set; }
    [DataMember(IsRequired = true)] public DateTime TimeUTC { get; set; }
    [DataMember(IsRequired = true)] public string Part { get; set; }
    [DataMember(IsRequired = true)] public int Quantity { get; set; }
  }

  // The following is only used for old decrement for backwards compatibility,
  // and shouldn't be used for anything else.
  [Serializable]
  public class JobAndPath : IEquatable<JobAndPath>
  {
    public readonly string UniqueStr;
    public readonly int Path;

    public JobAndPath(string unique, int path)
    {
      UniqueStr = unique;
      Path = path;
    }

    public bool Equals(JobAndPath other)
    {
      return (UniqueStr == other.UniqueStr && Path == other.Path);
    }
  }
}