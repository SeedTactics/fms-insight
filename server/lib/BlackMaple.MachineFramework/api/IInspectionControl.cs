/* Copyright (c) 2017, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace BlackMaple.MachineWatchInterface
{
  [Serializable, DataContract]
  public struct InspectCount
  {
    [DataMember(IsRequired = true)] public string Counter;
    [DataMember(IsRequired = true)] public int Value;
    [DataMember(IsRequired = true)] public DateTime LastUTC;
  }

  public interface IInspectionControl : IDisposable
  {
    ///Forces the given materialID to be inspected
    void ForceInspection(long materialID, string inspectionType);

    ///Forces the next piece that runs on the given location to be inspected
    void NextPieceInspection(MachineWatchInterface.PalletLocation palLoc, string inspType);

    //allow editing of counts.  Uses lists of InspectCount structure
    List<InspectCount> LoadInspectCounts();
    void SetInspectCounts(IEnumerable<InspectCount> countUpdates);
  }
}
