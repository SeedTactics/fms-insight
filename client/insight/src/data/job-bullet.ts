/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import { duration } from "moment";
import { Vector, Option } from "prelude-ts";

export interface DataPoint {
  readonly part: string;
  readonly completed: number;
  readonly completedCount: number;
  readonly totalPlan: number;
  readonly totalCount: number;
}

function displayJob(job: api.IInProcessJob, proc: number): DataPoint {
  const totalPlan = job.cyclesOnFirstProcess.reduce((a, b) => a + b, 0);
  const completed = job.completed ? job.completed[proc].reduce((a, b) => a + b, 0) : 0;

  const stops = job.procsAndPaths[proc].paths[0].stops;
  let cycleTime = duration();
  for (let i = 0; i < stops.length; i++) {
    const x = duration(stops[i].expectedCycleTime);
    cycleTime = cycleTime.add(x);
  }
  const cycleTimeHours = cycleTime.asHours();
  return {
    part: job.partName + "-" + (proc + 1).toString(),
    completed: cycleTimeHours * completed,
    completedCount: completed,
    totalPlan: cycleTimeHours * totalPlan,
    totalCount: totalPlan,
  };
}

export interface CompletedDataPoint extends DataPoint {
  readonly x: number;
  readonly y: number;
}

export interface DataPoints {
  readonly completedData: ReadonlyArray<CompletedDataPoint>;
  readonly planData: ReadonlyArray<{ x: number; y: number }>;
}

function vectorRange(start: number, count: number): Vector<number> {
  return Vector.unfoldRight(start, (x) =>
    x - start < count ? Option.of([x, x + 1] as [number, number]) : Option.none<[number, number]>()
  );
}

export function jobsToPoints(jobs: ReadonlyArray<Readonly<api.IInProcessJob>>): DataPoints {
  const points = Vector.ofIterable(jobs)
    .flatMap((j) => vectorRange(0, j.procsAndPaths.length).map((proc) => displayJob(j, proc)))
    .sortOn((pt) => pt.part)
    .reverse();
  const completedData = points
    .zipWithIndex()
    .map(([pt, i]) => ({ ...pt, x: pt.completed, y: i }))
    .toArray();
  const planData = points
    .zipWithIndex()
    .map(([pt, i]) => ({ x: pt.totalPlan, y: i }))
    .toArray();
  return { completedData, planData };
}
