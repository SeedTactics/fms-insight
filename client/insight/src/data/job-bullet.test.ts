/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { jobsToPoints } from "./job-bullet";
import * as api from "./api";

it("converts events to points", () => {
  const dummyHold: api.IJobHoldPattern = {
    userHold: false,
    reasonForUserHold: "",
    holdUnholdPattern: [],
    holdUnholdPatternStartUTC: new Date(),
    holdUnholdPatternRepeats: false,
  };
  const jobs: api.IInProcessJob[] = [
    {
      routeStartUTC: new Date(),
      routeEndUTC: new Date(),
      archived: false,
      copiedToSystem: true,
      partName: "part1",
      unique: "uniq1",
      priority: 10,
      manuallyCreated: false,
      createMarkingData: true,
      holdEntireJob: new api.JobHoldPattern(dummyHold),
      cyclesOnFirstProcess: [40, 42],
      completed: [[20, 21], [22]],
      procsAndPaths: [
        new api.ProcessInfo({
          paths: [
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal1"],
              load: [1],
              unload: [2],
              stops: [
                new api.JobMachiningStop({
                  stationNums: [1],
                  program: "progabc",
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "01:15:00",
                }),
              ],
              expectedLoadTime: "00:05:00",
              expectedUnloadTime: "00:04:00",
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            }),
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal2"],
              load: [1],
              unload: [2],
              stops: [
                new api.JobMachiningStop({
                  stationNums: [1],
                  program: "progabc",
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "01:15:00",
                }),
              ],
              expectedLoadTime: "00:05:00",
              expectedUnloadTime: "00:04:00",
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            }),
          ],
        }),
        new api.ProcessInfo({
          paths: [
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal1"],
              stops: [
                new api.JobMachiningStop({
                  stationNums: [1],
                  program: "progabc",
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "00:45:00",
                }),
              ],
              expectedLoadTime: "00:05:00",
              expectedUnloadTime: "00:04:00",
              load: [1],
              unload: [2],
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            }),
          ],
        }),
      ],
    },
  ];
  const points = jobsToPoints(jobs);
  expect(points).toMatchSnapshot("points for sample jobs");
});
