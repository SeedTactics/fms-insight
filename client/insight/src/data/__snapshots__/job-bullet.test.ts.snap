// Vitest Snapshot v1

exports[`converts events to points > points for sample jobs 1`] = `
{
  "jobs": [
    {
      "completed": 51.25,
      "completedCount": 41,
      "idx": 1,
      "part": "part1-1",
      "totalCount": 82,
      "totalPlan": 102.5,
      "workorders": "",
    },
    {
      "completed": 16.5,
      "completedCount": 22,
      "idx": 0,
      "part": "part1-2",
      "totalCount": 82,
      "totalPlan": 61.5,
      "workorders": "",
    },
  ],
  "longestPartName": 7,
}
`;
